atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionCacheValueStoreImpl.java

```diff
@@ -130,23 +134,49 @@ public TransactionCacheValueStore createWithFilteredSnapshot(CommitUpdate commit
                 .collect(Collectors.toSet());
     }
 
+    private void cacheHitInternal(TableReference table, Cell cell, CacheValue value) {
+        localUpdates.compute(CellReference.of(table, cell), (_unused, previousValue) -> {
+            if (previousValue != null) {
+                throw new SafeIllegalStateException(
+                        "Should not be attempting to record hits for keys that have been written to or read from the"
```

Comment:
This can race condition -- if we have a thread trying to read this cell, and another that wants to read or write simultaneously, we can read from the snapshot but then see that localUpdates have been updated in the meantime.
I guess it's reasonable to throw if the new value is different than ours, but surely we should just at least allow a parallel hit?

Reply:
All interactions with the TCVSI are synchronised

Reply:
(but without that, yes it can race)